pkg/infra/baseagent.go

```diff
@@ -93,8 +94,34 @@ func (a *Base) setHP(newHP int) {
 	a.hp = newHP
 }
 
-func (a *Base) updateHP(foodTaken float64) {
-	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
+func (a *Base) updateHP(foodTaken float64) { // first order system step answer. 63% of levelWidth obtained for tau food
+	switch {
+	// Strong level
+	case a.hp >= a.tower.healthInfo.StrongLevel:
+		tau := a.tower.healthInfo.FoodReqStrong // tau is the "time constant" of the first order system
+		shift := -tau * math.Log(1-(float64(a.hp)-float64(a.tower.healthInfo.StrongLevel))/a.tower.healthInfo.WidthStrong)
+		a.hp = a.tower.healthInfo.StrongLevel + int(a.tower.healthInfo.WidthStrong*(1-math.Pow(math.E, -(foodTaken+shift)/tau)))
+
+	// Healthy level
+	case a.hp >= a.tower.healthInfo.HealthyLevel:
+		tau := a.tower.healthInfo.FoodReqHealthy
+		shift := -tau * math.Log(1-(float64(a.hp)-float64(a.tower.healthInfo.HealthyLevel))/a.tower.healthInfo.WidthHealthy)
+		a.hp = a.tower.healthInfo.HealthyLevel + int(a.tower.healthInfo.WidthHealthy*(1-math.Pow(math.E, -(foodTaken+shift)/tau)))
+
+	// Weak Level
+	case a.hp >= a.tower.healthInfo.WeakLevel:
+		tau := a.tower.healthInfo.FoodReqWeak
+		shift := -tau * math.Log(1-(float64(a.hp)-float64(a.tower.healthInfo.WeakLevel))/a.tower.healthInfo.WidthWeak)
+		a.hp = a.tower.healthInfo.WeakLevel + int(a.tower.healthInfo.WidthWeak*(1-math.Pow(math.E, -(foodTaken+shift)/tau)))
+
+	// Critical Level - TODO: discuss its implementation
+	case a.hp >= 0:
+		if foodTaken >= a.tower.healthInfo.FoodReqCToW {
+			a.hp = int(a.tower.healthInfo.WeakLevel) - 1
+		} else {
+			a.hp = 1
+		}
+	}
```

Comment:
```suggestion
	// tau is the time constant of the first order system
	tau := 0.0
	base := 0
	width := 0.0

	switch {
	case a.hp >= a.tower.healthInfo.StrongLevel:
		tau = a.tower.healthInfo.FoodReqStrong 
		base = a.tower.healthInfo.StrongLevel
		width = a.tower.healthInfo.WidthStrong

	case a.hp >= a.tower.healthInfo.HealthyLevel:
		tau = a.tower.healthInfo.FoodReqHealthy
		base = a.tower.healthInfo.HealthyLevel
		width = a.tower.healthInfo.WidthHealthy

	case a.hp >= a.tower.healthInfo.WeakLevel:
		tau = a.tower.healthInfo.FoodReqWeak
		base = a.tower.healthInfo.WeakLevel
		width = a.tower.healthInfo.WidthWeak

	// Critical Level - TODO: discuss its implementation
	case a.hp >= 0:
		if foodTaken >= a.tower.healthInfo.FoodReqCToW {
			a.hp = int(a.tower.healthInfo.WeakLevel) - 1
		} else {
			a.hp = 1
		}
		return
	}

	a.hp = int(float64(base) + width - math.Pow(math.E, -foodTaken/tau) * float64(1 - a.hp + base))
```

Reply:
I worked through the formula you've given in the code for the hp calculation at the end... hopefully it's right. I've tried to simplify it so there's not as much computation required.

Reply:
+1 @jzzheng22 's implementation however could you make some clear documentation about the formula ahead of every case statement
`// HP = Level + Width * (1 - e^tau)`
Something like this on top maybe? And better documentation on what `Level`, `Width` and `tau` denote exactly.

Reply:
Good suggestion. I commit it but will slightly change the function line 129: you forgot the "shift". This shift might seem "magical" but it's not. I'll document its derivation and post it on slack. It is linked to the geogebra file + explanation I gave in slack.

Reply:
I'm not sure I agree. I didn't forget the shift: I substituted the formula for shift and simplified the final calculation.

Reply:
My bad, I didn't see the substitution. I think there is still a small mistake. Instead of:
`a.hp = int(float64(base) + width - math.Pow(math.E, -foodTaken/tau) * float64(1 - a.hp + base))` ,
it should be
`a.hp = int(float64(base) + width - math.Pow(math.E, -foodTaken/tau) * float64(width - a.hp + base))`.

I corrected this mistake and use your notation. I will also add the whole derivation in the wiki